lib/middleware/species.js

```diff
@@ -0,0 +1,17 @@
+const { pickBy, startsWith } = require('lodash');
+
+const process = () => (req, res, next) => {
+
+  req.form.values.modules.map(m => {
+    const specs = Object.values(pickBy(req.body, (value, key) => {
+      return startsWith(key, `module-${m}-species`);
+    })).filter(s => s !== '');
+    req.form.values[`module-${m}-species`] = specs;
+  });
+
+  next();
+};
+
+module.exports = {
+  process
+};
```

Comment:
This seems like the logic is in the wrong place - we're processing the form, then manually adding species after using this middleware. Firstly we should let the form do the processing by adding the pseudo fields to the schema before processing (in configure), then we should reset the schema to default after processing (in locals) - this way we don't lose the things we get for free with the for router such as validation, module cleaning etc.

Another point is this is a middleware which is extremely dependent on the environment in which it is called - it would be better to either export a function, or perhaps include a function in the schema file which takes modules and exports schema extended with the new fields

Reply:
done